vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PoolBase.java

```diff
@@ -97,21 +101,23 @@ public void getConnection(Handler<AsyncResult<SqlConnection>> handler) {
   }
 
   @Override
-  public <R> void schedule(CommandBase<R> cmd, Promise<R> promise) {
+  public <R> Future<R> schedule(ContextInternal context, CommandBase<R> cmd) {
+    Promise<R> promise = context.promise();
     Object metric;
     if (metrics != null) {
       metric = metrics.enqueueRequest();
     } else {
       metric = null;
     }
-    acquire(new CommandWaiter() {
+    acquire(context, new CommandWaiter() {
       @Override
       protected void onSuccess(Connection conn) {
         if (metrics != null) {
           metrics.dequeueRequest(metric);
         }
-        conn.schedule(cmd, promise);
-        promise.future().onComplete(ar -> {
+        conn.schedule(context, cmd)
+          .onComplete(promise)
+          .onComplete(v -> {
           // Use null promise instead
           conn.close(this, Promise.promise());
```

Comment:
I think we need to rework this because it won't work in pipelining mode. I find we have performance regression since https://github.com/eclipse-vertx/vertx-sql-client/pull/744, since all the pending commands will stuck instead of pipelining until the connection is returned back to the pool after the command completes. It can be reproduced by using sleep function and pipelining.

Reply:
oh really ? can we have a test that check pipelining ?

Reply:
I was actually thinking at something else for pipelining, to use the connection capacity to allow a connection to be borrowed multiple times fro mthe pool

Reply:
I'm first thinking how can we write a test that ensure pipelining works, any idea ?

Reply:
we can write a mini proxy server which can check if multiple query requests are sent to the server in the same round trip

Reply:
yes,but that requires to parse on the proxy the protocol

Reply:
well one thing we could do, is send 2 request with specific identifiers and look for those IDs and check that we have the 2 ids are written before the 2 ids are written back back on the response.

Reply:
@BillyYccc here is a tentative test that uses a proxy based that delays the packets sent to the database: https://github.com/eclipse-vertx/vertx-sql-client/pull/878

Reply:
I think one way we can make everyone happy is to use the connection capacity.

On the pool we define a pipelining limit (like for HTTP/1.x connection pool) and we set the initial capacity `C` of the connection to the limit, so the connection can be borrowed at most `C` times from the pool.

However this needs to be accommodated if a client wants an exclusive connection (i.e borrow with full capacity).

Reply:
> @BillyYccc here is a tentative test that uses a proxy that delays packets sent to the database: #878
> 
> This test fails on a pool but works on an un-pooled connection (not in the PR but checked by myself).

yes it works for unpooled connection since all scheduling commands are directly enqueued in the pending queue of the connection.

Reply:
> I think one way we can make everyone happy is to use the connection capacity.
> 
> On the pool we define a pipelining limit (like for HTTP/1.x connection pool) and we set the initial capacity `C` of the connection to the limit, so the connection can be borrowed at most `C` times from the pool.
> 
> However this needs to be accommodated if a client wants an exclusive connection (i.e borrow with full capacity).

yes we need to acquire exclusively for the stateful connection scene such as a transaction execution, we have to make sure the transactional commands are scheduled in the same connection.

Reply:
I have mixed feelings about implementing this as it does not seem to work well together, i.e having a pool policy that is both performant and fair does not seem possible, either the pool is fair and a request to an exclusive connection might block pipelining use or the pool might be unfair.

So I think the users should  have a choice to do between a pipelining pool and a pool that grants exclusive access to a connection.

Reply:
> So I would be inclined to have a pipelining limit=1 by default on `PoolOptions` and throw an exception if `getConnection` is called when limit>1.

Abandoing the usage of pipelining with `getConnection` is not good, because sometimes you might want to chain some operations in a reusable pooled connection without transactions but with pipelining enabled.
I wonder if we can abstract the fact and maintain two pools in the `Pool` interface, one is dedicated for `getConnection` use case and another one for one-shot executions. If the user calls `getConnection` then he gets a pooled connection and can do whatever he wants with the connection, commands like `transactions` will also be scheduled on the pool.

Reply:
> So I would be inclined to have a pipelining limit=1 by default on `PoolOptions` and throw an exception if `getConnection` is called when limit>1.

Abandoing the usage of pipelining with `getConnection` is not good, because sometimes you might want to chain some operations in a reusable pooled connection without transactions but with pipelining enabled, like cursor and batching.
I wonder if we can abstract the fact and maintain two pools in the `Pool` interface, one is dedicated for `getConnection` use case and another one for one-shot executions. If the user calls `getConnection` then he gets a pooled connection and can do whatever he wants with the connection, commands like `transactions` will also be scheduled on the pool.

Reply:
This would also introduce no breaking API changes for users but we might need some additional configuration params...

Reply:
So, we can support a non pipelined pool that uses pipelined connections, that is not a problem, i.e one can create a pool with pipelining limit = 1 on the pool but still use pipelining when borrowing a connection from the pool

Reply:
What we can have is a method that creates an `SqlClient` (so no getConnection method) that uses a pool of connections  that are pipelined, e.g

```java
SqlClient client = PgPool.pipelinedClient(poolOptions);
````

Reply:
or we can simply assume that `pool` methods will either throw an exception if pipeliningLimit > 1 or ignore this setting.

Reply:
@BillyYccc I implemented pool pipelining, in the way I removed the need to create a pooled connection when executing a simple command. Currently `PgPool` got a `client` method that will create a pipelined pool with a pipelining limit set to the connection pipelining limit.